torch/distributed/_sharded_tensor/ops/elementwise_ops.py

```diff
@@ -8,21 +8,18 @@
 )
 
 
-@sharded_op_impl(torch.nn.functional.gelu)
-def gelu(types, args=(), kwargs=None, pg=None):
-    """
-    Handles ``__torch_function__`` dispatch for the elementwise op
-    ``torch.nn.functional.gelu``.
-    This method computes on either a normal tensor or a sharded tensor.
-    """
-    input = args[0]
-    op = torch.nn.functional.gelu
-    # Validate types
-    if not isinstance(input, torch.Tensor) and not isinstance(input, ShardedTensor):
-        raise TypeError("input needs to be either torch.Tensor or ShardedTensor")
-    if isinstance(input, torch.Tensor):
-        return op(args[0])
-    else:
+def register_elementwise_op(op):
+    @sharded_op_impl(op)
+    def elementwise_op(types, args=(), kwargs=None, pg=None):
+        """
+        Handles ``__torch_function__`` dispatch for the elementwise op such
+        as ``torch.nn.functional.gelu`` or ``torch.nn.functional.relu``.
+        This method computes on either a normal tensor or a sharded tensor.
+        """
+        input = args[0]
+        # Validate types
+        if not isinstance(input, ShardedTensor):
+            raise TypeError("input needs to be a ShardedTensor")
         local_shards_new = []
         for local_shard in input.local_shards():
             local_shards_new.append(Shard(op(local_shard.tensor), local_shard.metadata))
```

Comment:
Lets add a TODO here to use the `from_local_tensor` API to avoid allgather overhead.

Reply:
Put in this issue: https://github.com/pytorch/pytorch/issues/72092